qiskit/qasm2/parse.py

```diff
@@ -0,0 +1,405 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Python-space bytecode interpreter for the output of the main Rust parser logic."""
+
+import dataclasses
+import math
+from typing import Iterable, Callable
+
+from qiskit.circuit import (
+    Barrier,
+    CircuitInstruction,
+    ClassicalRegister,
+    Delay,
+    Gate,
+    Instruction,
+    Measure,
+    QuantumCircuit,
+    QuantumRegister,
+    Qubit,
+    Reset,
+    library as lib,
+)
+
+# This is the same C-extension problems as described in the `__init__.py` disable near the
+# `_qasm2` import.
+from qiskit._qasm2 import (  # pylint: disable=no-name-in-module
+    OpCode,
+    UnaryOpCode,
+    BinaryOpCode,
+    CustomClassical,
+    ExprConstant,
+    ExprArgument,
+    ExprUnary,
+    ExprBinary,
+    ExprCustom,
+)
+from .exceptions import QASM2ParseError
+
+# Constructors of the form `*params -> Gate` for the special 'qelib1.inc' include.  This is
+# essentially a pre-parsed state for the file as given in the arXiv paper defining OQ2.
+QELIB1 = (
+    lib.U3Gate,
+    lib.U2Gate,
+    lib.U1Gate,
+    lib.CXGate,
+    # IGate in Terra < 0.24 is defined as a single-cycle delay, so is not strictly an identity.
+    # Instead we use a true no-op stand-in.
+    lambda: lib.UGate(0, 0, 0),
+    lib.XGate,
+    lib.YGate,
+    lib.ZGate,
+    lib.HGate,
+    lib.SGate,
+    lib.SdgGate,
+    lib.TGate,
+    lib.TdgGate,
+    lib.RXGate,
+    lib.RYGate,
+    lib.RZGate,
+    lib.CZGate,
+    lib.CYGate,
+    lib.CHGate,
+    lib.CCXGate,
+    lib.CRZGate,
+    lib.CU1Gate,
+    lib.CU3Gate,
+)
+
+
+@dataclasses.dataclass
```

Comment:
Can this be `frozen=True`?

Reply:
It can be if you think it's better? It's mostly just an ad-hoc structure, so I hadn't thought too much about anybody really storing instances or keeping multiple references to them.  The bytecode interpreter doesn't keep any references to instances of this class beyond the first couple of lines of the call.

Reply:
I find marking things as immutable/frozen incredibly useful when reading code. It removes a whole category of questions I have to pay attention to like "might this change over time?" I know it's nothing more than a container of data.

Beyond that philosophical motivation, `frozen=True` also causes `__hash__` to be implemented automatically (assuming `__eq__` isn't set to `False`). That doesn't matter here, but in general with Python, it's useful. 

I think `frozen=True` is a great default whenever writing dataclasses. I wish Python made it the default, but it would have confused a lot of people.

(It's part of why I love Rust, that things are immutable by default. Mutability is powerful, but requires cognitive overhead to keep track of it)

Reply:
I like immutable-by-default in languages that have proper support too, yeah.  Python doesn't really have it, though, so I find `frozen` and semi-hashability (this class will be hashable in some cases, but `constructor` might not be if it's a lambda, say) can sometimes mislead people into thinking things will work in general, since "frozen" can only apply to the dataclass itself and not its children.  The primitives guys have been having a similar-ish problem on `QuantumCircuit` - they wanted to produce a hashable key from `QuantumCircuit`, but it's pretty flaky because they can't guarantee that the contents of the tuple scalar value of `QuantumCircuit.data` will be recursively hashable.

That said, I'm happy to stick `frozen` on this - done in b9b1276ca.